keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+or by redefining what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- N examples of real world usage
+- N installs
+- More rigorous forms of testing e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least 2 releases between beta and
+GA/stable, since there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+#### Removing a deprecated flag
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality which deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+Currently, `kubelet --node-ip` has more-or-less no effect when using an
+external cloud provider. (It will require that the IP you pass is a
+known node IP, but it does not do anything else.) Assuming we change it
+to behave the same with external cloud providers as it does with legacy
+cloud providers and bare metal (ie, letting the user choose which node
+IP should be primary), then some users who were passing `--node-ip`
+before would need to stop passing it when upgrading, to avoid getting
+the new behavior. Perhaps this should warn and continue to use the old
+behavior during alpha, then warn and use the new behavior in beta.
+
+(Alternatively, since we need to extend `--node-ip` to support dual
+stack anyway, we could deprecate `--node-ip` and replace it with
+`--node-ips`, where the old `--node-ip` flag would continue to have the
+same semantics on external clouds until it went away.)
+
+Users on bare metal who upgrade and then convert their cluster from
+single-stack to dual-stack might need to switch back to single-stack
+first if they later needed to downgrade, due to the loss of some
+bare-metal-related dual-stack functionality when downgrading.
+
+External programs, scripts, etc that currently look at
+`Node.Status.Addresses` will eventually need to be updated to look at
+the new fields instead.
```

Comment:
.. I think (with this proposal) we can't make progress on dual-stack until we pass this "eventually need to be updated" point.  Is that correct?  If so, I'm concerned this KEP doesn't actually help us progress dual-stack.

(I like burning down `Node.Status.Addresses` and replacing it with something better defined.  I'm just considering the contextual motivation for this KEP, and I'm concerned that it actually hinders that goal rather than helps with it.)

Reply:
No, _external_ support for dual-stack doesn't matter. We expect that many external tools/extensions/plugins won't get around to adding support for dual-stack clusters until after dual stack goes GA. As long as we don't change anything that breaks those tools in _single-stack_ clusters, we're good.